### ISSUE TITLE
Prisma 4.8 produces conflicting typescript types for custom many to many relations

### ISSUE THREAD

**Issue by hayes** [prisma/prisma#17005]
### Bug description

prisma 4.8 is producing multiple typescript type definitions with the same name for a fairly common naming pattern for a custom many to many relation

### How to reproduce

1. clone https://github.com/hayes/prisma-4.8-generator-bug
2. run npm install
3. run npm geneate
4. run npm type

### Expected behavior

Type checking the client should not cause errors, types should all have unique names

### Prisma information

```prisma
datasource db {
  provider = "sqlite"
  url      = "file:./dev.db"
}

generator client {
  provider = "prisma-client-js"
  output   = "../client"
}

/// Model for Posts!
model Post {
  id Int @id @default(autoincrement())
  media PostMedia[]
}

model Media {
  id Int @id @default(autoincrement())
  posts PostMedia[]
}

model PostMedia {
  id      Int   @id @default(autoincrement())
  post    Post  @relation(fields: [postId], references: [id])
  media   Media @relation(fields: [mediaId], references: [id])
  postId  Int
  mediaId Int
  @@unique([postId, mediaId])
}
```

```ts
// Add your code using Prisma Client
```

### Environment & setup

- OS: macOS
- Database: sqlite
- Node.js version: 16.17

### Prisma Version

```
prisma                  : 4.8.0
@prisma/client          : 4.8.0
Current platform        : darwin-arm64
Query Engine (Node-API) : libquery-engine d6e67a83f971b175a593ccc12e15c4a757f93ffe (at node_modules/.pnpm/@prisma+engines@4.8.0/node_modules/@prisma/engines/libquery_engine-darwin-arm64.dylib.node)
Migration Engine        : migration-engine-cli d6e67a83f971b175a593ccc12e15c4a757f93ffe (at node_modules/.pnpm/@prisma+engines@4.8.0/node_modules/@prisma/engines/migration-engine-darwin-arm64)
Introspection Engine    : introspection-core d6e67a83f971b175a593ccc12e15c4a757f93ffe (at node_modules/.pnpm/@prisma+engines@4.8.0/node_modules/@prisma/engines/introspection-engine-darwin-arm64)
Format Binary           : prisma-fmt d6e67a83f971b175a593ccc12e15c4a757f93ffe (at node_modules/.pnpm/@prisma+engines@4.8.0/node_modules/@prisma/engines/prisma-fmt-darwin-arm64)
Format Wasm             : @prisma/prisma-fmt-wasm 4.8.0-61.d6e67a83f971b175a593ccc12e15c4a757f93ffe
Default Engines Hash    : d6e67a83f971b175a593ccc12e15c4a757f93ffe
Studio                  : 0.479.0
```

**Comment by jkomyno** [comment]
I can confirm this issue happens with `prisma@4.8.0`, but not with `prisma@4.7.1`.

**Comment by Jolg42** [comment]
This looks related to https://github.com/prisma/prisma/pull/16844 @SevInf 

Released in our internal version `4.8.0-dev.70`
https://buildkite.com/prisma/release-prisma-typescript/builds/7780#01852ecf-dbdc-4a4d-ac8d-c8ea399483a9

`4.8.0-dev.69` I can see 
```ts
  /**
   * PostMedia without action
   */
  export type PostMediaArgs = {
    /**
     * Select specific fields to fetch from the PostMedia
     * 
    **/
    select?: PostMediaSelect | null
    /**
     * Choose, which related nodes to fetch as well.
     * 
    **/
    include?: PostMediaInclude | null
  }
```
 and since `4.8.0-dev.70`
```
// line 1907

  /**
   * Post.media
   */
  export type PostMediaArgs = {
    /**
     * Select specific fields to fetch from the PostMedia
     * 
    **/
    select?: PostMediaSelect | null
    /**
     * Choose, which related nodes to fetch as well.
     * 
    **/
    include?: PostMediaInclude | null
    where?: PostMediaWhereInput
    orderBy?: Enumerable<PostMediaOrderByWithRelationInput>
    cursor?: PostMediaWhereUniqueInput
    take?: number
    skip?: number
    distinct?: Enumerable<PostMediaScalarFieldEnum>
  }

// line 3967

  /**
   * PostMedia without action
   */
  export type PostMediaArgs = {
    /**
     * Select specific fields to fetch from the PostMedia
     * 
    **/
    select?: PostMediaSelect | null
    /**
     * Choose, which related nodes to fetch as well.
     * 
    **/
    include?: PostMediaInclude | null
  }
```

**Comment by SevInf** [comment]
Note: fix will be published as 4.8.1 version today

**Comment by serg06** [comment]
Is this only for explicit relations or for implicit ones as well?

**Comment by HugoxSaraiva** [comment]
Hi, can we reopen this issue, or should I open a new one?

I'm getting the same error on prisma v4.16.2, but with a different schema definition.
I've created a repo to reproduce this error on 4.16.2:
https://github.com/HugoxSaraiva/prisma-4.16-generator-bug

Node version: v16.19.1

Prisma version:
```
prisma                  : 4.16.2
@prisma/client          : 4.16.2
Current platform        : darwin-arm64
Query Engine (Node-API) : libquery-engine 4bc8b6e1b66cb932731fb1bdbbc550d1e010de81 (at node_modules/@prisma/engines/libquery_engine-darwin-arm64.dylib.node)
Migration Engine        : migration-engine-cli 4bc8b6e1b66cb932731fb1bdbbc550d1e010de81 (at node_modules/@prisma/engines/migration-engine-darwin-arm64)
Format Wasm             : @prisma/prisma-fmt-wasm 4.16.1-1.4bc8b6e1b66cb932731fb1bdbbc550d1e010de81
Default Engines Hash    : 4bc8b6e1b66cb932731fb1bdbbc550d1e010de81
Studio                  : 0.484.0
```

**Comment by SevInf** [comment]
@HugoxSaraiva I took a look at your repository and this specific issue have been fixed in 5.1, I'd suggest updating.

In generally, the chance that the issue you are experiencing is exactly the same as the one that was fixed 7 month ago are pretty low, even if the error message seems similar. So, I'd always advice to check latest Prisma version first and if you still have the problem then open a new issue, please.
